### ISSUE TITLE
Is it possible to use UVR via command line?

### ISSUE THREAD

**Issue by czhou** [Anjok07/ultimatevocalremovergui#359]
like:
`uvr --model=htdemucs_ft --model=another_model --input=input.mp3 --output-dir=/path/to/ouput
`

**Comment by Anjok07** [comment]
Not yet. It's on my list of things to do. I will announce it when it's ready.

**Comment by techzealot** [comment]
cli will be great!!!

**Comment by jarredou** [comment]
Yeah, I would also appreciate a command line feature !

**Comment by leochras** [comment]
Working on a bash script right now, which downloads any video, splits the stems and transcribes lyrics. As spleeter is not really usable anymore, being able to split stems via cli would be a huge improvement!

**Comment by crexlight** [comment]
> Not yet. It's on my list of things to do. It'll likely be awhile since I've been extremely busy.

Any progress information?

**Comment by beveradb** [comment]
Hey folks, in case this is helpful to anyone...

I have a simple karaoke creation use case and I already know the MDX models trained by @Anjok07 work for UVR work really well for this (especially `UVR_MDXNET_KARA_2`), but I wanted to be able to easily run the separation from the command line or from my other python hobby projects.

So, I created this little python package, splitting out the minimal parts from the UVR repo in order to do a simple separation of a WAV file into the primary & secondary stems WAV files.

If your use case is simple or similar to mine (e.g. you don't need any non-MDX models, you don't need more than 2 stems, or secondary processing, advanced settings, etc. - you just need a song to be split into instrumental and vocals - then this may be useful to you!

It can either be installed using pip and run from the CLI, or imported as a python package and used in some other project (example in README).

https://github.com/karaokenerds/python-audio-separator

To use on the CLI, you can simply install it from PyPI:
```
pip install audio-separator
```

and run with a single CLI param specifying a WAV file to separate, e.g.
```
audio-separator input.wav
```

and it should output 2 WAV files in the current directory after about 20 seconds.

it defaults to using the `UVR_MDXNET_KARA_2` model, but you can use any other model just by specifying it by name with the `--model_name` CLI arg/param.

It downloads both the model metadata and model file itself on demand to `/tmp/audio-separator-models`, but you can also override this with the `--model_file_dir` arg/param. 

To be clear: what I've created here is really just a refactoring and simplification of one part of Anjok's code - he absolutely deserves credit if this is useful to anyone! (https://www.buymeacoffee.com/uvr5 👀) 

and uh, I hope this is ok @Anjok07 😅  - I did consider forking UVR and trying to raise a PR here to add a CLI mode, but I kinda wanted a super minimal thing which just worked without any of the additional dependencies etc. in the full-fat UVR repo.

Anyway, i'm off to sleep now but wanted to share this in case it helps someone 🙂

**Comment by daswer123** [comment]
@beveradb Bro thanks, I was just making myself a bot in tg and wanted to add this functionality.

**Comment by seanghay** [comment]
A simple CLI version of UVR: https://github.com/seanghay/uvr

**Comment by venkat-sf** [comment]
@Anjok07 Any chance of CLI happening soon?

**Comment by ziyuang** [comment]
@beveradb do I understand correctly that ensembling is not included?

**Comment by ziyuang** [comment]
@beveradb can I do ensembling?

**Comment by beveradb** [comment]
Sorry @ziyuang  nobody has implemented Ensemble yet - PRs welcome!
https://github.com/nomadkaraoke/python-audio-separator/issues/12

**Comment by jarredou** [comment]
You can ensemble already separated audio with https://github.com/ZFTurbo/Music-Source-Separation-Training and its ensemble.py script

**Comment by ziyuang** [comment]
> You can ensemble already separated audio with https://github.com/ZFTurbo/Music-Source-Separation-Training and its ensemble.py script

This looks comprehensive. Will need some time for figure out the best argument combination for vocal vs. instrument with my VRAM constraint (6GB by the way).

**Comment by ziyuang** [comment]
> Will need some time for figure out the best argument combination for vocal vs. instrument

OK at least the [algorithm description](https://mvsep.com/algorithms/10?lang=en) mentions that the best models for ensembling are Mel Roformer, BS Roformer and SCNet XL. The ensembling methods and weights are not known though.

**Comment by jarredou** [comment]
I've just shared my own script to find best weights for ensembles https://github.com/jarredou/weight_finder_v2
It's only using "avg_wave" type of ensemble currently but it can be edited to allow more ensemble types, more details on the ensemble types here https://github.com/ZFTurbo/Music-Source-Separation-Training/blob/main/docs/ensemble.md

**Comment by ziyuang** [comment]
> I've just shared my own script to find best weights for ensembles https://github.com/jarredou/weight_finder_v2 It's only using "avg_wave" type of ensemble currently but it can be edited to allow more ensemble types, more details on the ensemble types here https://github.com/ZFTurbo/Music-Source-Separation-Training/blob/main/docs/ensemble.md

Great stuff. Do you mind sharing your finding?

**Comment by jarredou** [comment]
@ziyuang I've used that script to find these optimal weights for these three models (that are available in UVR). I haven't made tests with other model combinations, as it was made for my [MDX23-colab-fork](https://github.com/jarredou/MVSEP-MDX23-Colab_v2). The evaluations are based on multisong evaluation dataset from MVSEP.
![Image](https://github.com/user-attachments/assets/01eba096-a761-4bac-a444-5eb8fe7a6ad7)

**Comment by tremby** [comment]
For anyone as unobservant as me:

With no modifications, this GUI *does* accept multiple inputs. From the "select input" dialog it's possible to select multiple files, and they'll be processed sequentially.

That's the main thing I wanted a CLI for. Maybe you were looking for the same thing. Hopefully I saved you the hour I just spent feeding files in one at a time.

**Comment by chyinan** [comment]
Hey everyone! 👋

Massive thanks to @Anjok07 for the incredible work on UVR, and to @beveradb for the great `audio-separator` package which paved the way for CLI usage.

For those looking for a **fully Dockerized, Headless solution** designed specifically for **server/cloud deployments** (e.g., RunPod, AutoDL, or headless Linux servers) without dealing with dependency hell, I've built:

### [uvr-headless-runner](https://github.com/chyinan/uvr-headless-runner)

**Key Differences / Features:**
* 🐳 **Native Docker Support**: Just pull and run. No Python environment setup required. Perfect for batch processing on GPUs.
* 🚀 **Full Architecture Support**: Supports **BS-Roformer**, **SCNet**, **MDX23C**, **Demucs v4**, and **VR Arch** out of the box.
* 🎛️ **Official Logic**: Designed to mimic the official UVR5 processing logic as closely as possible for high-fidelity results.
* **Smart Caching**: Automated model downloading and caching system.

It’s intended for power users who need to process TBs of audio or integrate UVR into a backend service. Hope this helps anyone looking for a robust server-side alternative!

PRs and Feedback are welcome!

**Comment by beveradb** [comment]
> For those looking for a **fully Dockerized, Headless solution** designed specifically for **server/cloud deployments** (e.g., RunPod, AutoDL, or headless Linux servers) without dealing with dependency hell, I've built:
> ### [uvr-headless-runner](https://github.com/chyinan/uvr-headless-runner)

Maybe I'm missing something @chyinan but how does this differ from the Dockerfile (including one optimized for Runpod) I already provide, or the Modal.com cloud deployment infra in the audio-separator repo? See:
- https://github.com/nomadkaraoke/python-audio-separator?tab=readme-ov-file#-docker
- https://github.com/nomadkaraoke/python-audio-separator/blob/main/audio_separator/remote/README.md

**Comment by chyinan** [comment]
> > For those looking for a **fully Dockerized, Headless solution** designed specifically for **server/cloud deployments** (e.g., RunPod, AutoDL, or headless Linux servers) without dealing with dependency hell, I've built:
> > ### [uvr-headless-runner](https://github.com/chyinan/uvr-headless-runner)
> 
> Maybe I'm missing something [@chyinan](https://github.com/chyinan) but how does this differ from the Dockerfile (including one optimized for Runpod) I already provide, or the Modal.com cloud deployment infra in the audio-separator repo? See:
> 
> * https://github.com/nomadkaraoke/python-audio-separator?tab=readme-ov-file#-docker
> * https://github.com/nomadkaraoke/python-audio-separator/blob/main/audio_separator/remote/README.md

Hi @beveradb! Thanks for checking in and for the great work on audio-separator!

You are absolutely right—your repo does have excellent Docker/Cloud support (I missed emphasizing that distinction clearly, my apologies!).

The main difference I'm aiming for with uvr-headless-runner is strictly adhering to the official UVR5 processing pipeline logic.

audio-separator is a fantastic re-implementation/library that makes these models accessible and modular for developers.

uvr-headless-runner is designed more as a direct headless wrapper around the original UVR codebase.

The goal is to ensure that users who prototype specific parameter combinations (e.g., TTA, specific aggressiveness curves) in the official GUI can expect bit-exact (or as close as possible) results when moving to headless batch processing, without worrying about potential implementation differences in the inference loop.

Basically, yours is the go-to Library for integration, while mine is a specialized Runner for those who want the exact "UVR5 experience" on a server. Hope that clears it up! :)
